### PR TITLE
Support promises for Node.js

### PR DESCRIPTION
**Pull request by jasonpet**
Add support for promises.
JSHint cleanup as well.

### CODE REVIEW

**Comment by jasonpet** [comment]
whisk.done() and whisk.async() will be removed for nodejs 6 as part of this pr since they are not necessary now that promises are supported.  However, with the change being made to make nodejs 6 the default in https://github.com/openwhisk/openwhisk/pull/738 all async js actions (such as pipe.js etc.....) installed by the install scripts will fail when run.  This is because the install scripts do not specify the kind so nodejs 6 will be used once it becomes the default.  As part of this pr I will update the install scripts to specify the kind to use nodejs for all of the async js actions.  I will will open a new pr to handle converting our existing async js actions to use promises.

**Comment by sjfink** [comment]
I suggest we `deprecate` async and done rather than remove them.   That is, just leave them there for now and document that people should start using Promises instead.

**Comment by jasonpet** [comment]
pg build 212

**Comment by jasonpet** [comment]
@psuter - assigning to you since you own #78 as well.  I made some simple tweaks to the commit from #78 so it would apply to the current code base.

**Comment by jasonpet** [comment]
@csantanapr - I will open a new issue to document that promises can now be used instead of async and done as Steve had suggested.  Where should we document this?  I see doc references for using async in references.md and actions.md.  Is this where the promises documentation should go?  Should we remove the references to async in the doc?

**Comment by jasonpet** [comment]
I am going to add a simple sample action showing how an action can return a promise and add a test that uses the sample.

**Comment by csantanapr** [comment]
actions.md add simple example with promises, 
in references.md mark async and done as deprecated and provide more in depth details about promises (i.e. how to use bluebird, q, etc..) @bjustin-ibm might have better ideas on how to structure this.

**Comment by psuter** [comment]
Looks good to me, will merge when we have the example in docs+tests. Feel free to use additional commits as the blame shouldn't fall on the original author when the tests fail :)  As a point of curiosity, since the change is in the common container, does this mean we can use, e.g. bluebird packaged with webpack, and work with promises in the Node 0.9 container?

**Comment by bjustin-ibm** [comment]
> (i.e. how to use bluebird, q, etc..)

It looks like I misunderstood the Promises documentation regarding the use of other libraries, and you do not get toll-free bridging for other promises libraries.

@psuter You are free to use other promise packages in the Node 0.12 runtime, but in the end, you will need to utilize `whisk.async()` and `whisk.done()` to inform the OW runtime what you are up to.  Similarly, in the Node 6 runtime, you will ultimately need to return an ES6 Promise, even if you are using another promise library in your code.

**Comment by jasonpet** [comment]
Native promises are supported in the Node 0.12.14 container so there is no need to use bluebird in that container.

**Comment by jasonpet** [comment]
I added a sample for Promises and a test for the sample as well.  I also found that the existing tests I had created to test Node.js 0.12.14 and Node.js 6 are both testing Node.js 6 now that the default Node has changed.  I updated the tests to specify the --kind for testing Node.js 0.12.14.

**Comment by jasonpet** [comment]
Created issue https://github.com/openwhisk/openwhisk/issues/825 to handle the documentation for Promises.

File: core/nodejsActionBase/runner.js [hunk 1]
```diff
@@ -71,13 +71,20 @@ function NodeActionRunner(message, whisk) {
...
+            var self = this;
+            Promise.resolve(result) //Non-promises/undefined instantly resolve.
+                .then(function(resolvedResult) {
+                    // This happens, e.g. if you just have "return;"
+                    if (typeof resolvedResult === 'undefined') {
+                        resolvedResult = {};
+                    }
+                    closeAndReturn(resolvedResult);
+                }, function (error) {
+                        self.whisk.error(error);
```
Comment (rabbah) [inline core/nodejsActionBase/runner.js]: is there a test case to cover the error case that reaches this path?

File: tests/src/actionContainers/NodeJsActionContainerTests.scala [hunk 2]
```diff
@@ -357,4 +357,50 @@ class NodeJsActionContainerTests extends FlatSpec with Matchers {
         filtered(out).trim shouldBe empty
         filtered(err).trim shouldBe empty
     }
+
+    it should "support resolved promises" in {
+      val (out, err) = withNodeJsContainer { c =>
+      val code = """
```
Comment (rabbah) [inline tests/src/actionContainers/NodeJsActionContainerTests.scala]: is line 364 deliberately aligned one character to the left?

File: core/nodejsActionBase/runner.js (hunk 1 above)
Comment (jasonpet) [inline core/nodejsActionBase/runner.js]: yes, there is an action container test in this PR that rejects the Promise and will hit this path.

**Comment by jasonpet** [comment]
pg build 251 ran with latest updates.  @psuter - this should be ready to merge now.

**Comment by psuter** [comment]
Glad we now have this in!

**Comment by sjfink** [comment]
@jasonpet in addition to the official documentation, how about a blog post on this?

**Comment by Samze** [comment]
Great to see this made it in :)

**Pull request merged**